components/channel/channel.h

```diff
@@ -0,0 +1,78 @@
+#ifndef DATADOG_PHP_CHANNEL_H
+#define DATADOG_PHP_CHANNEL_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+typedef struct datadog_php_receiver_s datadog_php_receiver;
+typedef struct datadog_php_sender_s datadog_php_sender;
+typedef struct datadog_php_channel_s datadog_php_channel;
+typedef struct datadog_php_channel_impl_s datadog_php_channel_impl;
+
+struct datadog_php_receiver_s {
+    /**
+     * Receives on item, waiting for a send to the channel if necessary.
+     * However, it may return empty-handed and return false, such as if there are
+     * no senders left, or the timeout is reached, or if it is interrupted.
+     */
+    bool (*recv)(datadog_php_receiver *self, void **data, uint64_t timeout_nanos);
+
+    /**
+     * Destructs the receiver.
+     */
+    void (*dtor)(datadog_php_receiver *self);
+
+    // private:
+    datadog_php_channel_impl *channel;
+};
+
+struct datadog_php_sender_s {
+    /**
+     * Sends data through the channel. May fail, in which case the caller is
+     * responsible for dtor'ing the data if applicable. Does not wait for the
+     * data to be received, only enqueued.
+     */
+    bool (*send)(datadog_php_sender *self, void *data);
+
+    /**
+     * Clones the sender. Can fail if there are too many senders already, in
+     * which false is returned and the caller should not dtor the clone.
+     */
+    bool (*clone)(datadog_php_sender *self, datadog_php_sender *clone);
+
+    /**
+     * Destructs the sender. Must be done on each successful clone as well.
+     */
+    void (*dtor)(datadog_php_sender *self);
+
+    // private:
+    datadog_php_channel_impl *channel;
+};
+
+/**
+ * datadog_php_channel is a bounded, multiple-producer, single-consumer channel
+ * suitable for inter-thread communication.
+ *
+ * The caller is responsible for any additional cleanup of contents; it does
```

Comment:
I'm concerned that what we have here is a buffered channel - producers don't synchronize with consumers, so the producer may free what is "buffered" and yet to be consumed. But making it the callers responsibility to ensure this doesn't happen makes the component hard to use.

Can you talk a little about your reasoning here ?

Reply:
Ah, "caller" was not a good choice there: it should say "user" instead.

An example: I malloc an object that represents a stack sample, and then send it through the channel. If it fails to enqueue, then the caller frees it. If it succeeds, then the receiver frees it. A conditional lifetime is not great but seemed better than the alternatives because we're in C. Feel free to propose something different if you think something that might be better.

Reply:
I'll do thoughts and may push some more spec/tests

Reply:
I changed `caller` to `user`, which is at least a slightly more accurate description.